cmd/influx-tools/importer/importer.go

```diff
@@ -0,0 +1,186 @@
+package importer
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strconv"
+	"time"
+
+	"github.com/influxdata/influxdb/cmd/influx-tools/server"
+	"github.com/influxdata/influxdb/services/meta"
+	"github.com/influxdata/influxdb/tsdb"
+	"github.com/influxdata/influxdb/tsdb/engine/tsm1"
+	"go.uber.org/zap"
+)
+
+type Config struct {
+	Database string
+	Data     tsdb.Config
+}
+
+type Importer struct {
+	MetaClient server.MetaClient
+	db         string
+	dataDir    string
+
+	rpi       *meta.RetentionPolicyInfo
+	log       *zap.Logger
+	sh        *shardWriter
+	sfile     *tsdb.SeriesFile
+	sw        *seriesWriter
+	buildTsi  bool
+	seriesBuf []byte
+}
+
+const seriesBatchSize = 1000
+
+func NewImporter(client server.MetaClient, db string, dataDir string, buildTsi bool, log *zap.Logger) *Importer {
+	i := &Importer{MetaClient: client, db: db, dataDir: dataDir, buildTsi: buildTsi, log: log}
+	if !buildTsi {
+		i.seriesBuf = make([]byte, 0, 2048)
+	}
+	return i
+}
+
+func (i *Importer) Close() error {
+	defer i.CloseShardGroup()
+	err := i.closeSeriesFile()
+	if err != nil {
+		return err
+	}
+	return i.CloseShardGroup()
+}
+
+func (i *Importer) CreateDatabase(rp *meta.RetentionPolicySpec, replace bool) error {
+	var rpi *meta.RetentionPolicyInfo
+	dbInfo := i.MetaClient.Database(i.db)
+	if dbInfo == nil {
+		return i.createDatabaseWithRetentionPolicy(rp)
+	}
+
+	rpi, err := i.MetaClient.RetentionPolicy(i.db, rp.Name)
+	if err != nil {
+		return err
+	}
+
+	updateRp := (rpi != nil) && replace && ((rpi.Duration != *rp.Duration) || (rpi.ShardGroupDuration != rp.ShardGroupDuration))
+	if updateRp {
+		err = i.updateRetentionPolicy(rpi, rp)
+		if err != nil {
+			return err
+		}
+	} else {
+		_, err = i.MetaClient.CreateRetentionPolicy(i.db, rp, false)
+		if err != nil {
+			return err
+		}
+	}
+
+	return i.createDatabaseWithRetentionPolicy(rp)
+}
+
+func (i *Importer) updateRetentionPolicy(oldRpi *meta.RetentionPolicyInfo, newRp *meta.RetentionPolicySpec) error {
+	for _, shardGroup := range oldRpi.ShardGroups {
+		err := i.MetaClient.DeleteShardGroup(i.db, newRp.Name, shardGroup.ID)
```

Comment:
when you delete a shard group from the metadata store, you will also need to remove the shard directories on disk. An example of this can be found in the `retention` service as follows:

* Delete shard group from meta store [here](https://github.com/influxdata/influxdb/blob/12b1ecc40ee64bf79999a3bcdd34a7f457c0e5ed/services/retention/service.go#L108)
* Delete shard group from disk [here](https://github.com/influxdata/influxdb/blob/12b1ecc40ee64bf79999a3bcdd34a7f457c0e5ed/services/retention/service.go#L134)